tests/pytests/unit/utils/test_pycrypto.py

```diff
@@ -0,0 +1,190 @@
+
+    ret = salt.utils.pycrypto.gen_hash(
+        crypt_salt=expected["badsalt"], password=passwd, algorithm=algorithm
+    )
+    assert ret != expected["hashed"]
+
+    ret = salt.utils.pycrypto.gen_hash(
+        crypt_salt=None, password=passwd, algorithm=algorithm
+    )
+    assert ret != expected["hashed"]
+
+
+@pytest.mark.skipif(not salt.utils.pycrypto.HAS_CRYPT, reason="crypt not available")
+def test_gen_hash_crypt_no_arguments():
+    # Assert it works without arguments passed
+    assert salt.utils.pycrypto.gen_hash() is not None
+
+
+@pytest.mark.skipif(not salt.utils.pycrypto.HAS_CRYPT, reason="crypt not available")
+def test_gen_hash_crypt_default_algorithm():
+    # Assert it works without algorithm passed
+    default_algorithm = salt.utils.pycrypto.crypt.methods[0].name.lower()
+    expected = expecteds[default_algorithm]
+    ret = salt.utils.pycrypto.gen_hash(crypt_salt=expected["salt"], password=passwd)
+    assert ret == expected["hashed"]
+
+
+@pytest.mark.skipif(not salt.utils.pycrypto.HAS_PASSLIB, reason="passlib not available")
+@patch("salt.utils.pycrypto.methods", {})
+@patch("salt.utils.pycrypto.HAS_CRYPT", False)
+@pytest.mark.parametrize(
+    "algorithm, expected",
+    [
+        ("sha512", expecteds["sha512"]),
+        ("sha256", expecteds["sha256"]),
+        ("blowfish", expecteds["blowfish"]),
+        ("md5", expecteds["md5"]),
+        ("crypt", expecteds["crypt"]),
+    ],
+)
+def test_gen_hash_passlib(algorithm, expected):
+    """
+    Test gen_hash with passlib
+    """
+    ret = salt.utils.pycrypto.gen_hash(
+        crypt_salt=expected["salt"], password=passwd, algorithm=algorithm
+    )
+    assert ret == expected["hashed"]
+
+    ret = salt.utils.pycrypto.gen_hash(
+        crypt_salt=expected["badsalt"], password=passwd, algorithm=algorithm
+    )
+    assert ret != expected["hashed"]
+
+    ret = salt.utils.pycrypto.gen_hash(
+        crypt_salt=None, password=passwd, algorithm=algorithm
+    )
+    assert ret != expected["hashed"]
+
+
+def test_gen_hash_passlib_no_arguments():
+    # Assert it works without arguments passed
+    assert salt.utils.pycrypto.gen_hash() is not None
+
+
+def test_gen_hash_passlib_default_algorithm():
+    # Assert it works without algorithm passed
+    default_algorithm = salt.utils.pycrypto.known_methods[0]
+    expected = expecteds[default_algorithm]
+    if default_algorithm in expected:
+        ret = salt.utils.pycrypto.gen_hash(crypt_salt=expected["salt"], password=passwd)
+        assert ret == expected["hashed"]
+
+
+@patch("salt.utils.pycrypto.HAS_CRYPT", False)
+@patch("salt.utils.pycrypto.HAS_PASSLIB", False)
+def test_gen_hash_no_lib():
+    """
+    test gen_hash with no crypt library available
+    """
+    with pytest.raises(SaltInvocationError):
+        salt.utils.pycrypto.gen_hash()
+
+
+@patch("salt.utils.pycrypto.HAS_CRYPT", True)
+@patch("salt.utils.pycrypto.methods", {"crypt": None})
+@patch("salt.utils.pycrypto.HAS_PASSLIB", True)
```

Comment:
These and all other patch decorators

Reply:
Yeah, sorry sent my message a little too quickly before updating :)

Reply:
Yeah, I also noticed that I had already mentioned the patch decorators previously, sorry.

Reply:
Just checked... believe i got them all now.